sys/timer.c

```diff
@@ -0,0 +1,113 @@
+#define KL_LOG KL_TIME
+#include <timer.h>
+#include <mutex.h>
+#include <errno.h>
+#include <klog.h>
+#include <interrupt.h>
+
+static mtx_t timers_mtx = MTX_INITIALIZER(MTX_DEF);
+static timer_list_t timers = TAILQ_HEAD_INITIALIZER(timers);
+
+#define TMF_ACTIVE 0x1000
+#define TMF_INITIALIZED 0x2000
+#define TMF_ALLOCATED 0x4000
+#define TMF_REGISTERED 0x8000
+
+#define is_active(tm) ((tm)->tm_flags & TMF_ACTIVE)
+#define is_initialized(tm) ((tm)->tm_flags & TMF_INITIALIZED)
+#define is_allocated(tm) ((tm)->tm_flags & TMF_ALLOCATED)
+#define is_registered(tm) ((tm)->tm_flags & TMF_REGISTERED)
+
+int tm_register(timer_t *tm) {
+  if (is_registered(tm))
+    return EBUSY;
+
+  WITH_MTX_LOCK (&timers_mtx) {
+    TAILQ_INSERT_TAIL(&timers, tm, tm_link);
+    tm->tm_flags &= TMF_TYPEMASK;
+    tm->tm_flags |= TMF_REGISTERED;
+  }
+
+  klog("Registered '%s' timer.", tm->tm_name);
+  return 0;
+}
+
+int tm_deregister(timer_t *tm) {
+  assert(is_registered(tm));
+
+  if (is_allocated(tm))
+    return EBUSY;
+
+  WITH_MTX_LOCK (&timers_mtx) {
+    TAILQ_REMOVE(&timers, tm, tm_link);
+    tm->tm_flags &= TMF_TYPEMASK;
+  }
+
+  klog("Unregistered '%s' timer.", tm->tm_name);
+  return 0;
+}
+
+timer_t *tm_alloc(const char *name, unsigned flags) {
+  timer_t *tm = NULL;
```

Comment:
`name` argument of `tm_alloc` is unused.  Also, alloc is a bit misleading word here. Consider renaming, e.g. to `tm_lock`.

Reply:
I renamed it to `tm_reserve`. Please note that you missed the fact that `tm_release` (previously `tm_free`) is not implemented.

Reply:
Ok. I fixed timer reserve by name.